### PR TITLE
Port most vLLM kernels to ROCm

### PR DESCRIPTION
**Pull request by pcmoritz**
This ports most vLLM kernels to ROCm (with the exception of the `quantization_ops` which is not critical to run some of the models).

If you have a working ROCm installation, you can compile this with `python setup.py develop` or `python setup.py install` (on my local machine, `pip install -e .` is NOT working, not sure if that is generic or specific to my setup).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -64,66 +60,6 @@ def get_torch_arch_list() -> Set[str]:
     return set(arch_list)
 
 
-# First, check the TORCH_CUDA_ARCH_LIST environment variable.
```
Comment (pcmoritz) [inline setup.py:67]: Maybe we can put all of this code into a function? Then we can just skip it if `torch.version.hip` is set :)

If you have a different preference, let me know!

File: csrc/activation_kernels.cu [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #include <torch/extension.h>
 #include <ATen/cuda/CUDAContext.h>
 
+#include "cuda_compat.h"
```
Comment (ardfork) [inline csrc/activation_kernels.cu:4]: That file doesn't exist.

Comment (pcmoritz) [reply csrc/activation_kernels.cu:4]: Sorry, added! Is there anything else missing that prevents you from building this? If yes please let me know :)

**Comment by ardfork** [comment]
It does build, but I don't think I have enough VRAM to run any unquantized model. Also, note that it require ROCm 5.7, amd_hip_bf16.h didn't exist before.

Anyway, doesn't it have a hard dependency on xFormers? Just trying to call vllm.entrypoints.openai.api_server throw `ModuleNotFoundError: No module named 'xformers'`. All the xFormers part need to have a different path if it's not installed.

**Comment by pcmoritz** [comment]
The first step here will be to get one of the smaller models like OPT-125M working, which doesn't need AWQ. For dealing with `xformers` my plan is to work with https://github.com/pcmoritz/vllm-public/tree/flash-attn for the time being :)

**Comment by casper-hansen** [comment]
@pcmoritz Curious if you would mind looking into porting the AWQ quantization kernels to ROCm too? Would be a benefit to everyone running quantized models.

**Comment by pcmoritz** [comment]
@casper-hansen It depends on how the rest of the port goes and how the performance numbers for the rest looks like. I hope we can get some help from AMD to port them since it is likely more involved :)

**Comment by pcmoritz** [comment]
@WoosukKwon Can you have a look at the PR? All the vLLM layers are working now on AMD hardware except the ones that depend on xformers (with the following patch that I didn't want to merge). Make sure to have a closer look at and test the refactoring in setup.py since I haven't tested it on nvidia hardware :)
```
(base) anyscale@phx2-pc2-017:/tmp/vllm-public$ git diff
diff --git a/tests/kernels/test_activation.py b/tests/kernels/test_activation.py
index 0b3ad0a..44aa944 100644
--- a/tests/kernels/test_activation.py
+++ b/tests/kernels/test_activation.py
@@ -72,4 +72,4 @@ def test_gelu_fast(
     out = torch.empty(num_tokens, d, dtype=dtype, device="cuda")
     activation_ops.gelu_fast(out, x)
     ref_out = get_activation("gelu_fast")(x)
-    assert torch.allclose(out, ref_out, atol=1e-5, rtol=1e-5)
+    assert torch.allclose(out, ref_out, atol=1e-3, rtol=1e-3)
diff --git a/tests/kernels/test_attention.py b/tests/kernels/test_attention.py
index 31d78dd..82ecc09 100644
--- a/tests/kernels/test_attention.py
+++ b/tests/kernels/test_attention.py
@@ -12,7 +12,7 @@ from vllm.utils import get_max_shared_memory_bytes
 FLOAT32_BYTES = torch.finfo(torch.float).bits // 8
 # This will change depending on the compute capability.
 # - 512 as a buffer
-MAX_SEQ_LEN = get_max_shared_memory_bytes() // FLOAT32_BYTES - 512
+MAX_SEQ_LEN = 8192
 NUM_BLOCKS = 128  # Arbitrary values for testing
 PARTITION_SIZE = 512
 
diff --git a/tests/kernels/test_pos_encoding.py b/tests/kernels/test_pos_encoding.py
index d660417..0f66a9e 100644
--- a/tests/kernels/test_pos_encoding.py
+++ b/tests/kernels/test_pos_encoding.py
@@ -170,5 +170,5 @@ def test_rotary_embedding(
     ref_key = ref_key.view(num_tokens, num_heads * head_size)
 
     # Compare the results.
-    assert torch.allclose(out_query, ref_query, atol=1e-5, rtol=1e-5)
-    assert torch.allclose(out_key, ref_key, atol=1e-5, rtol=1e-5)
+    assert torch.allclose(out_query, ref_query, atol=1e-3, rtol=1e-3)
+    assert torch.allclose(out_key, ref_key, atol=1e-3, rtol=1e-3)
(base) anyscale@phx2-pc2-017:/tmp/vllm-public$ ls tests/kernels/
conftest.py  __pycache__  test_activation.py  test_attention.py  test_cache.py  test_layernorm.py  test_pos_encoding.py
(base) anyscale@phx2-pc2-017:/tmp/vllm-public$ python -m pytest -s tests/kernels/test_activation.py 
WARNING 10-16 23:45:51 ray_utils.py:35] Failed to import Ray with ModuleNotFoundError("No module named 'msgpack'"). For distributed inference, please install Ray with `pip install ray pandas pyarrow`.
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.11.4, pytest-7.4.2, pluggy-1.0.0
rootdir: /tmp/vllm-public
collected 108 items                                                                                                                                                                                                                        

tests/kernels/test_activation.py ............................................................................................................

=========================================================================================================== 108 passed in 0.50s ============================================================================================================
(base) anyscale@phx2-pc2-017:/tmp/vllm-public$ python -m pytest -s tests/kernels/test_cache.py
WARNING 10-16 23:46:02 ray_utils.py:35] Failed to import Ray with ModuleNotFoundError("No module named 'msgpack'"). For distributed inference, please install Ray with `pip install ray pandas pyarrow`.
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.11.4, pytest-7.4.2, pluggy-1.0.0
rootdir: /tmp/vllm-public
collected 270 items                                                                                                                                                                                                                        

tests/kernels/test_cache.py ..............................................................................................................................................................................................................................................................................

=========================================================================================================== 270 passed in 6.21s ============================================================================================================
(base) anyscale@phx2-pc2-017:/tmp/vllm-public$ python -m pytest -s tests/kernels/test_layernorm.py
WARNING 10-16 23:46:15 ray_utils.py:35] Failed to import Ray with ModuleNotFoundError("No module named 'msgpack'"). For distributed inference, please install Ray with `pip install ray pandas pyarrow`.
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.11.4, pytest-7.4.2, pluggy-1.0.0
rootdir: /tmp/vllm-public
collected 45 items                                                                                                                                                                                                                         

tests/kernels/test_layernorm.py .............................................

============================================================================================================ 45 passed in 0.44s ============================================================================================================
(base) anyscale@phx2-pc2-017:/tmp/vllm-public$ python -m pytest -s tests/kernels/test_pos_encoding.py 
WARNING 10-16 23:46:21 ray_utils.py:35] Failed to import Ray with ModuleNotFoundError("No module named 'msgpack'"). For distributed inference, please install Ray with `pip install ray pandas pyarrow`.
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.11.4, pytest-7.4.2, pluggy-1.0.0
rootdir: /tmp/vllm-public
collected 864 items                                                                                                                                                                                                                        

tests/kernels/test_pos_encoding.py ................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

=========================================================================================================== 864 passed in 3.77s ============================================================================================================
(base) anyscale@phx2-pc2-017:/tmp/vllm-public$ python -m pytest -s tests/kernels/test_attention.py 
WARNING 10-16 23:46:37 ray_utils.py:35] Failed to import Ray with ModuleNotFoundError("No module named 'msgpack'"). For distributed inference, please install Ray with `pip install ray pandas pyarrow`.
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.11.4, pytest-7.4.2, pluggy-1.0.0
rootdir: /tmp/vllm-public
collecting ... Deterministic: False
Performance Mode: True
collected 324 items                                                                                                                                                                                                                        

tests/kernels/test_attention.py ................................................................................................................................................................................................................................................................................................FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
```
(the failures are the ones from xformers modules not being available)

**Comment by WoosukKwon** [comment]
Hi @pcmoritz, huge thanks for the great work! I will definitely take a look. As for the FlashAttention and xformers, we will probably be able to collaborate with the AI teams at AMD.

**Comment by WoosukKwon** [comment]
@pcmoritz @casper-hansen Speaking of the AWQ kernels, I believe they are for temporary use. We plan to implement much faster kernels (probably using Triton) in the near future.

**Comment by jingqiao** [comment]
Thanks for the great work @pcmoritz ! @WoosukKwon !

FYI - As to FlashAttention, MosaicML posted a [blog](https://www.mosaicml.com/blog/amd-mi250) on LLM training with AMD's ML250 via ROCmSoftwarePlatform's relevant effort on FlashAttention. It might be worth having a test with it

_From their blog_
- Start with a ROCm 5.4  Docker image: `[rocm/dev-ubuntu-20.04:5.4.3-complete](https://hub.docker.com/r/rocm/dev-ubuntu-20.04/tags)`
- Install PyTorch for ROCm 5.4: `pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/rocm5.4.2`
- Install ROCm-based FlashAttention following [the instructions here](https://github.com/ROCmSoftwarePlatform/flash-attention/tree/flash_attention_for_rocm2#amd-gpurocm-support).

**Comment by sabreshao** [comment]
@pcmoritz @jingqiao @WoosukKwon I am the leader of [flash-attention](https://github.com/ROCmSoftwarePlatform/flash-attention) in AMD.
Please reach out me if you need a newer flash attention with newer ROCm.

**Comment by WoosukKwon** [comment]
@pcmoritz Thanks again for the great work! I have a quick question: How did you port the CUDA kernels to HIP? Did you use the AMD's hipify script or just do it manually?

**Comment by pcmoritz** [comment]
I did it manually (the hipify script is actually called under the hood by the compiler to translate the file, which mostly replaces the `cuda*` APIs with `hip*` APIs). In my first iteration, I called hipify on the files but then you get a bunch of copied files and that would be hard to maintain. Later I figured out that hipify is called automatically under the hood, and this PR just contains the stuff that hipify can't convert :)

**Comment by pcmoritz** [comment]
@iAmir97 converted the script to AMD assembler instructions in https://github.com/pcmoritz/vllm-public/pull/1/files, let me try if I can cherrypick the commits into this PR and if the tests still pass, then maybe that's the better solution since the performance might be better :)

**Comment by pcmoritz** [comment]
The tests are passing, so I changed the code.

I'll be at skycamp tomorrow @WoosukKwon, if you want and have time, we can chat / hack some more on this :)

**Comment by fsx950223** [comment]
When I tried flash_attn branch, I got NameError: name 'BlockDiagonalCausalMask' is not defined.

**Comment by fsx950223** [comment]
Does asm work as expect?

**Comment by pcmoritz** [comment]
@fsx950223 It is passing the unit tests for the layers, so seems to be working for me. On the `BlockDiagonalCausalMask` error -- there are some more changes that need to be made to get flash attention working, see https://github.com/vllm-project/vllm/pull/1314. It is not working fully for me yet but I'm also planning to spend some more time on it -- let me know if you'd like to help, it should be possible to get it working :)

**Comment by pcmoritz** [comment]
@sabreshao I'm planning to work more on flash attention integration, currently I'm working with the latest master in https://github.com/ROCmSoftwarePlatform/flash-attention, let me know if you think using a different code base will make it more likely to succeed :) Currently the focus is on correctness and not speed yet :)

**Comment by fsx950223** [comment]
> @fsx950223 It is passing the unit tests for the layers, so seems to be working for me. On the `BlockDiagonalCausalMask` error -- there are some more changes that need to be made to get flash attention working, see #1314. It is not working fully for me yet but I'm also planning to spend some more time on it -- let me know if you'd like to help, it should be possible to get it working :)

Could you share the development environment with me (maybe a docker file)? I met some issues on my local environment. And what's the type of GPU you are using MI250 or MI300?

**Comment by jingqiao** [comment]
@pcmoritz @WoosukKwon @sabreshao 

Below is a workable solution to enable Flash Attention for vLLM per my testing. Pls just treat it as a reference, and I am sure you can make further optimizations.

To be specific, it is to follow the instruction at https://github.com/ROCmSoftwarePlatform/flash-attention except that, instead of using ROCm 5.4, we can use ROCm5.7 (already supported by pytorch nightly), the same version @pcmoritz used above.

To be specific, one can replace the following [code](https://github.com/vllm-project/vllm/blob/d189170b6c5a143e493c3f5cb7e8c976e8da62c7/vllm/model_executor/layers/attention.py#L107) in vLLM
```
        out = xops.memory_efficient_attention_forward(
            query.unsqueeze(0),
            key.unsqueeze(0),
            value.unsqueeze(0),
            attn_bias=input_metadata.attn_bias,
            p=0.0,
            scale=self.scale,
        )
```
with the following code (similar to the [code here](https://github.com/ROCmSoftwarePlatform/flash-attention/blob/444e15a7b2136058372d10a0d9714a8b8ac3efec/benchmarks/benchmark_flash_attention.py#L54))
```
        from einops import repeat
        from flash_attn.bert_padding import unpad_input, pad_input
        from flash_attn.flash_attn_interface import flash_attn_unpadded_qkvpacked_func

        # Use Flash Attention (with ROCm support)
        batch_size = len(input_metadata.prompt_lens)
        seq_len = input_metadata.max_prompt_len

        lengths = torch.tensor(input_metadata.prompt_lens, device=query.device).view(-1,1)
        attention_mask_bool = repeat(
            torch.arange(seq_len, device=query.device),
            's -> b s', b=batch_size
        ) < lengths

        qkv = torch.stack([query, key, value], dim=1).view(
            batch_size, seq_len, 3, self.num_heads, self.head_size
        )
        qkv_unpad, indices, cu_seqlens, max_seqlen_in_batch = unpad_input(qkv, attention_mask_bool)

        out_unpad = flash_attn_unpadded_qkvpacked_func(
            qkv_unpad, cu_seqlens, max_seqlen_in_batch, dropout_p=0.0, causal=True
        )
        out = pad_input(out_unpad, indices, batch_size, seq_len).view(
                  -1, self.num_heads, self.head_size
        )
```

**Comment by WoosukKwon** [comment]
Hi @pcmoritz, I got the following error when building vLLM:
```
In file included from csrc/attention/attention_kernels.hip:23:
    In file included from csrc/attention/attention_dtypes_hip.h:7:
    In file included from csrc/attention/dtype_bfloat16_hip.cuh:29:
    /opt/rocm-5.6.0/include/hip/hip_bf16.h:30:10: fatal error: 'hip/amd_detail/amd_hip_bf16.h' file not found
    #include <hip/amd_detail/amd_hip_bf16.h>
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    1 error generated when compiling for gfx1030.
    error: command '/opt/rocm-5.6.0/bin/hipcc' failed with exit code 1
```
Do you happen to know how to resolve this?

I'm using PyTorch 2.1.0 with ROCm 5.6.0. My local ROCm version is also 5.6.0.

**Comment by pcmoritz** [comment]
This header is only available with ROCm 5.7 unfortunately (somebody else
ran into the same problem)

**Comment by WoosukKwon** [comment]
@pcmoritz I see. Then, which version of PyTorch did you use?

**Comment by pcmoritz** [comment]
The nightly one that comes with ROCm 5.7 support

**Comment by fsx950223** [comment]
I optimized the performance on ROCM, with the patch, the performance improved about 10%.
diff --git a/csrc/attention/attention_kernels.cu b/csrc/attention/attention_kernels.cu
index babd15b..4937f55 100644
--- a/csrc/attention/attention_kernels.cu
+++ b/csrc/attention/attention_kernels.cu
@@ -23,7 +23,11 @@
 
 #include <algorithm>
 
-#define WARP_SIZE 32
+#ifndef USE_ROCM
+  #define DEFAULT_NUM_THREADS 128
+#else
+  #define DEFAULT_NUM_THREADS 256
+#endif
 #define MAX(a, b) ((a) > (b) ? (a) : (b))
 #define MIN(a, b) ((a) < (b) ? (a) : (b))
 #define DIVIDE_ROUND_UP(a, b) (((a) + (b) - 1) / (b))
@@ -39,7 +43,7 @@ inline __device__ float block_sum(float* red_smem, float sum) {
 
   // Compute the sum per warp.
 #pragma unroll
-  for (int mask = WARP_SIZE / 2; mask >= 1; mask /= 2) {
+  for (int mask = WARP_SIZE / 2; mask >= 1; mask >>= 1) {
     sum += VLLM_SHFL_XOR_SYNC(sum, mask);
   }
 
@@ -58,7 +62,7 @@ inline __device__ float block_sum(float* red_smem, float sum) {
 
   // Parallel reduction inside the warp.
 #pragma unroll
-  for (int mask = NUM_WARPS / 2; mask >= 1; mask /= 2) {
+  for (int mask = NUM_WARPS / 2; mask >= 1; mask >>= 1) {
     sum += VLLM_SHFL_XOR_SYNC(sum, mask);
   }
 
@@ -114,8 +118,6 @@ __device__ void paged_attention_kernel(
   const int num_tokens = end_token_idx - start_token_idx;
 
   constexpr int THREAD_GROUP_SIZE = MAX(WARP_SIZE / BLOCK_SIZE, 1);
-  constexpr int NUM_THREAD_GROUPS = NUM_THREADS / THREAD_GROUP_SIZE; // Note: This assumes THREAD_GROUP_SIZE divides NUM_THREADS
-  assert(NUM_THREADS % THREAD_GROUP_SIZE == 0);
   constexpr int NUM_TOKENS_PER_THREAD_GROUP = DIVIDE_ROUND_UP(BLOCK_SIZE, WARP_SIZE);
   constexpr int NUM_WARPS = NUM_THREADS / WARP_SIZE;
   const int thread_idx = threadIdx.x;
@@ -149,11 +151,11 @@ __device__ void paged_attention_kernel(
   // th vectors of the query, and so on.
   // NOTE(woosuk): Because q is split from a qkv tensor, it may not be contiguous.
   const scalar_t* q_ptr = q + seq_idx * q_stride + head_idx * HEAD_SIZE;
-  __shared__ Q_vec q_vecs[THREAD_GROUP_SIZE][NUM_VECS_PER_THREAD];
+  Q_vec q_vecs[NUM_VECS_PER_THREAD];
 #pragma unroll
-  for (int i = thread_group_idx; i < NUM_VECS_PER_THREAD; i += NUM_THREAD_GROUPS) {
+  for (int i = 0; i < NUM_VECS_PER_THREAD; i++) {
     const int vec_idx = thread_group_offset + i * THREAD_GROUP_SIZE;
-    q_vecs[thread_group_offset][i] = *reinterpret_cast<const Q_vec*>(q_ptr + vec_idx * VEC_SIZE);
+    q_vecs[i] = *reinterpret_cast<const Q_vec*>(q_ptr + vec_idx * VEC_SIZE);
   }
   __syncthreads(); // TODO(naed90): possible speedup if this is replaced with a memory wall right before we use q_vecs
 
@@ -200,7 +202,7 @@ __device__ void paged_attention_kernel(
 
       // Compute dot product.
       // This includes a reduction across the threads in the same thread group.
-      float qk = scale * Qk_dot<scalar_t, THREAD_GROUP_SIZE>::dot(q_vecs[thread_group_offset], k_vecs);
+      float qk = scale * Qk_dot<scalar_t, THREAD_GROUP_SIZE>::dot(q_vecs, k_vecs);
       // Add the ALiBi bias if slopes are given.
       qk += (alibi_slope != 0) ? alibi_slope * (token_idx - context_len + 1) : 0;
 
@@ -231,7 +233,7 @@ __device__ void paged_attention_kernel(
   // Get the max qk value for the sequence.
   qk_max = lane < NUM_WARPS ? red_smem[lane] : -FLT_MAX;
 #pragma unroll
-  for (int mask = NUM_WARPS / 2; mask >= 1; mask /= 2) {
+  for (int mask = NUM_WARPS / 2; mask >= 1; mask >>= 1) {
     qk_max = fmaxf(qk_max, VLLM_SHFL_XOR_SYNC(qk_max, mask));
   }
   // Broadcast the max qk value to all threads.
@@ -319,7 +321,7 @@ __device__ void paged_attention_kernel(
   for (int i = 0; i < NUM_ROWS_PER_THREAD; i++) {
     float acc = accs[i];
 #pragma unroll
-    for (int mask = NUM_V_VECS_PER_ROW / 2; mask >= 1; mask /= 2) {
+    for (int mask = NUM_V_VECS_PER_ROW / 2; mask >= 1; mask >>= 1) {
       acc += VLLM_SHFL_XOR_SYNC(acc, mask);
     }
     accs[i] = acc;
@@ -332,8 +334,8 @@ __device__ void paged_attention_kernel(
   // Perform reduction across warps.
   float* out_smem = reinterpret_cast<float*>(shared_mem);
 #pragma unroll
-  for (int i = NUM_WARPS; i > 1; i /= 2) {
-    int mid = i / 2;
+  for (int i = NUM_WARPS; i > 1; i >>= 1) {
+    int mid = i >> 1;
     // Upper warps write to shared memory.
     if (warp_idx >= mid && warp_idx < i) {
       float* dst = &out_smem[(warp_idx - mid) * HEAD_SIZE];
@@ -485,7 +487,7 @@ __global__ void paged_attention_v2_reduce_kernel(
   // Get the global max logit.
   // Reduce within the warp.
 #pragma unroll
-  for (int mask = WARP_SIZE / 2; mask >= 1; mask /= 2) {
+  for (int mask = WARP_SIZE / 2; mask >= 1; mask >>= 1) {
     max_logit = fmaxf(max_logit, VLLM_SHFL_XOR_SYNC(max_logit, mask));
   }
   if (lane == 0) {
@@ -495,7 +497,7 @@ __global__ void paged_attention_v2_reduce_kernel(
   // Reduce across warps.
   max_logit = lane < NUM_WARPS ? red_smem[lane] : -FLT_MAX;
 #pragma unroll
-  for (int mask = NUM_WARPS / 2; mask >= 1; mask /= 2) {
+  for (int mask = NUM_WARPS / 2; mask >= 1; mask >>= 1) {
     max_logit = fmaxf(max_logit, VLLM_SHFL_XOR_SYNC(max_logit, mask));
   }
   // Broadcast the max value to all threads.
@@ -533,9 +535,9 @@ __global__ void paged_attention_v2_reduce_kernel(
 } // namespace vllm
 
 #define LAUNCH_PAGED_ATTENTION_V1(HEAD_SIZE)                                                  \
-  cudaFuncSetAttribute(                                                                       \
+  hipFuncSetAttribute(                                                                       \
     (void*)vllm::paged_attention_v1_kernel<T, HEAD_SIZE, BLOCK_SIZE, NUM_THREADS>,            \
-    cudaFuncAttributeMaxDynamicSharedMemorySize, shared_mem_size);                            \
+    hipFuncAttributeMaxDynamicSharedMemorySize, shared_mem_size);                            \
   vllm::paged_attention_v1_kernel<T, HEAD_SIZE, BLOCK_SIZE, NUM_THREADS>                      \
   <<<grid, block, shared_mem_size, stream>>>(                                                 \
     out_ptr,                                                                                  \
@@ -556,7 +558,7 @@ __global__ void paged_attention_v2_reduce_kernel(
 template<
   typename T,
   int BLOCK_SIZE,
-  int NUM_THREADS = 128>
+  int NUM_THREADS = DEFAULT_NUM_THREADS>
 void paged_attention_v1_launcher(
   torch::Tensor& out,
   torch::Tensor& query,
@@ -715,7 +717,7 @@ void paged_attention_v1(
 template<
   typename T,
   int BLOCK_SIZE,
-  int NUM_THREADS = 128,
+  int NUM_THREADS = DEFAULT_NUM_THREADS,
   int PARTITION_SIZE = 512>
 void paged_attention_v2_launcher(
   torch::Tensor& out,
diff --git a/csrc/attention/attention_utils.cuh b/csrc/attention/attention_utils.cuh
index ff64c4b..1987ad4 100644
--- a/csrc/attention/attention_utils.cuh
+++ b/csrc/attention/attention_utils.cuh
@@ -39,7 +39,7 @@ inline __device__ float qk_dot_(const Vec (&q)[N], const Vec (&k)[N]) {
   // Finalize the reduction across lanes.
   float qk = sum(qk_vec);
 #pragma unroll
-  for (int mask = THREAD_GROUP_SIZE / 2; mask >= 1; mask /= 2) {
+  for (int mask = THREAD_GROUP_SIZE / 2; mask >= 1; mask >>= 1) {
     qk += VLLM_SHFL_XOR_SYNC(qk, mask);
   }
   return qk;
diff --git a/csrc/cuda_compat.h b/csrc/cuda_compat.h
index 8991462..f56cbd0 100644
--- a/csrc/cuda_compat.h
+++ b/csrc/cuda_compat.h
@@ -17,3 +17,10 @@
 #else
   #define VLLM_SHFL_SYNC(var, src_lane) __shfl(var, src_lane)
 #endif
+
+
+#ifndef USE_ROCM
+  #define WARP_SIZE 32
+#else
+  #define WARP_SIZE 64
+#endif
diff --git a/csrc/reduction_utils.cuh b/csrc/reduction_utils.cuh
index b95ccef..563b776 100644
--- a/csrc/reduction_utils.cuh
+++ b/csrc/reduction_utils.cuh
@@ -19,12 +19,13 @@
 
 #include "cuda_compat.h"
 
+
 namespace vllm {
 
 template<typename T>
 __inline__ __device__ T warpReduceSum(T val) {
 #pragma unroll
-  for (int mask = 16; mask > 0; mask >>= 1)
+  for (int mask = WARP_SIZE / 2; mask > 0; mask >>= 1)
     val += VLLM_SHFL_XOR_SYNC(val, mask);
   return val;
 }
@@ -32,9 +33,15 @@ __inline__ __device__ T warpReduceSum(T val) {
 /* Calculate the sum of all elements in a block */
 template<typename T>
 __inline__ __device__ T blockReduceSum(T val) {
-  static __shared__ T shared[32];
+  static __shared__ T shared[WARP_SIZE];
+
+#ifndef USE_ROCM
   int lane = threadIdx.x & 0x1f;
   int wid = threadIdx.x >> 5;
+#else
+  int lane = threadIdx.x & 0x3f;
+  int wid = threadIdx.x >> 6;
+#endif
 
   val = warpReduceSum<T>(val);
 
@@ -45,7 +52,7 @@ __inline__ __device__ T blockReduceSum(T val) {
 
   // Modify from blockDim.x << 5 to blockDim.x / 32. to prevent
   // blockDim.x is not divided by 32
-  val = (threadIdx.x < (blockDim.x / 32.f)) ? shared[lane] : (T)(0.0f);
+  val = (threadIdx.x < (blockDim.x / float(WARP_SIZE))) ? shared[lane] : (T)(0.0f);
   val = warpReduceSum<T>(val);
   return val;
 }

**Comment by fxmarty** [comment]
@pcmoritz Thank you for the port. Are you aware of a way to compile against RoCm 5.6?

**Comment by ehartford** [comment]
need to support 5.6 because thats what pytorch supports now

**Comment by fxmarty** [comment]
@ehartford Yes it would be nice. So far I have been using nightly and it works quite nicely!

**Comment by ehartford** [comment]
Ok I figured out how to install both versions of rocm so that I can use 5.7 with vllm and 5.6 with pytorch

**Comment by ehartford** [comment]
what's preventing this from being merged?

**Comment by ehartford** [comment]
```
(vllm) eric@quixi1:~/vllm-public$ python setup.py install
Traceback (most recent call last):
  File "/home/eric/vllm-public/setup.py", line 28, in <module>
    raise RuntimeError(
RuntimeError: Cannot find CUDA_HOME. CUDA must be available to build the package.
```

**Comment by iAmir97** [comment]
@WoosukKwon Do you have any idea when the changing of AWQ kernels will happen? planning to port them to rocm.

**Comment by pcmoritz** [comment]
@ehartford Are you on the right branch? In this branch, this should be handled (https://github.com/vllm-project/vllm/pull/1313/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R27) -- your error seems to indicate that the `raise RuntimeError` line is line 28, which is not the case in this branch :)

**Comment by WoosukKwon** [comment]
Closed as we merged #1836 which is a superset of this PR. @pcmoritz Thanks for the amazing work!

**Pull request closed without merging**